covid/impl/chainbinom_simulate.py

```diff
@@ -19,18 +19,33 @@ def propagate_fn(t, state):
                                          -tf.reduce_sum(rate_matrix, axis=-1))
         # Calculate Markov transition probability matrix
         markov_transition = tf.linalg.expm(rate_matrix*time_step)
-        # Sample new state
-        new_state = tfd.Multinomial(total_count=state,
-                                    probs=markov_transition).sample()
-        new_state = tf.reduce_sum(new_state, axis=-1)
+        num_states = markov_transition.shape[-1]
+        prev_prob = tf.zeros_like(markov_transition[..., :, 0])
+        counts = tf.zeros(markov_transition.shape[:-1].as_list() + [0],
+                          dtype=markov_transition.dtype)
+        total_count = state
+        # This for loop is ok because there are (currently) only 4 states (SEIR)
+        # and we're only actually creating work for 3 of them. Even for as many
+        # as a ~10 states it should probably be fine, just increasing the size
+        # of the graph a bit.
+        for i in range(num_states - 1):
+          binom = tfd.Binomial(
+              total_count=total_count,
+              probs=markov_transition[..., :, i] / (1. - prev_prob))
+          sample = binom.sample()
+          counts = tf.concat([counts, sample[..., tf.newaxis]], axis=-1)
+          total_count -= sample
+          prev_prob = binom.probs
```

Comment:
Is this probability update actually correct?  Mentally stepping through the loop for desired probs = [1/4, 1/4, 1/4, 1/4], I think we get

Start of iteration 0:
prev_prob = 0
binom.probs = 1/4
Start of iteration 1:
prev_prob = 1/4
binom.probs = 1/3  # So far so good
Start of iteration 2:
prev_prob = 1/3  # Um...
binom.probs = 3/8  # Definitely not good

Should this instead be
```
probs_here = markov_transition[..., :, i]
tfd.Binomial(..., probs=probs_here / (1. - prev_prob)
prev_prob += probs_here
```
?

Also, probably a good idea to clip the probs passed to Binomial with `tf.clip_by_value(probs, 0, 1)`, because the rejection sampler will busy-loop if it gets a prob of, say, 1.000001.

Reply:
I believe you're right, thanks for checking my work Alexey...Having this properly implemented and unit tested in TFP will be ideal :) Meanwhile I'll prep a PR for use here that corrects my errors.

Reply:
created #4